### PR TITLE
support scalable multi-workflows

### PR DESCRIPTION
**Pull request by damianxu88**
the change to support multi-workflows returned from orchard payload generation, this is tested successfully in QA for both old and new `--array` spade generate command

### CODE REVIEW

**Comment by schowsf** [comment]
Does spade need any change to go with this?  I am thinking not, assuming spadeCli already generates Orchard json payloads with line breaks, and the `OrchardClient` code on `range result` treats line breaks as field separators.

**Comment by damianxu88** [comment]
@schowsf no spade changes needed, the orchard payload results are piped into stdout with line break in between

**Review by schowsf — APPROVED** [review]
LGTM

**Review by schowsf — APPROVED** [review]

File: service/scheduler.go [hunk 1]
```diff
@@ -80,29 +80,37 @@ func (s *Scheduler) lockAndRun(db *gorm.DB, wf table.Workflow) {
...
-			fmt.Println(err)
-			notifyOwner(wf, err)
+		for _, orchardID := range orchardIDs {
+			err = client.Activate(orchardID)
+			if err != nil {
+				fmt.Println(err)
+				notifyOwner(wf, err)
+				scheduleStatus[orchardID] = "error"
+			} else {
+				scheduleStatus[orchardID] = "activated"
```
Comment (anniezhang77) [inline service/scheduler.go:96]: do we still want to activate the other workflows when there is error in one of the workflows?

Comment (damianxu88) [reply service/scheduler.go:96]: @anniezhang77 yes, the next functionality to add is rollback when encountering errors

File: orchard/orchard_runner.go [hunk 2]
```diff
@@ -94,25 +94,30 @@ func cmdOutput(cmd *exec.Cmd) ([]byte, []byte, error) {
...
 	if err != nil {
 		combinedOutput := fmt.Sprintf("%s\n%s", output, string(stderr))
-		return "", fmt.Errorf("exec command %v has error: %w: %s", command, err, combinedOutput)
+		return []string{}, fmt.Errorf("exec command %v has error: %w: %s", command, err, combinedOutput)
 	}
-	return output, nil
+	var outputs []string
+	err = json.Unmarshal([]byte(output), &outputs)
+	if err != nil {
+		return []string{}, fmt.Errorf("marshall error %w", err)
```
Comment (realstraw) [inline orchard/orchard_runner.go:118]: let's just wrap this with an array and return the original content, so that it's backwards compatible

Comment (damianxu88) [reply orchard/orchard_runner.go:118]: @realstraw still return the original `output`? so `return output, fmt.Errorf("marshall error %w", err)`?

Comment (damianxu88) [reply orchard/orchard_runner.go:118]: @realstraw gotcha, wrap if error to make it backward compatible

File: service/scheduler.go (hunk 1 above)
Comment (damianxu88) [reply service/scheduler.go]: rollback added

**Review by schowsf — APPROVED** [review]

**Comment by damianxu88** [comment]
@realstraw this is ready for another review and I've tested this in QA environment

File: service/scheduler.go [hunk 3]
```diff
@@ -51,6 +78,49 @@ func (s *Scheduler) scheduleWorkflows(db *gorm.DB) {
 	}
 }
 
+func (s *Scheduler) createActivateWorkflow(
+	client *orchard.OrchardRestClient,
+	wf table.Workflow,
+) map[string]string {
+	scheduleStatus := make(map[string]string)
```
[the remark this answers is not in the document]
Comment (damianxu88) [reply service/scheduler.go:85]: sure, let me change

File: service/scheduler.go [hunk 4]
```diff
@@ -51,6 +78,68 @@ func (s *Scheduler) scheduleWorkflows(db *gorm.DB) {
...
+		fmt.Printf("[error] error creating workflow: %s\n", err)
+		notifyOwner(wf, err)
+		statuses = s.deleteWorkflows(client, createdIDs, statuses)
+	} else {
+		for _, createdID := range createdIDs {
+			err = client.Activate(createdID)
+			if err != nil {
+				fmt.Printf("[error] error activating workflow: %s\n", err)
+				notifyOwner(wf, err)
+				statuses = s.cancelWorkflows(client, statuses)
```
Comment (realstraw) [inline service/scheduler.go:134]: you can break at this point already, no?

Comment (damianxu88) [reply service/scheduler.go]: @realstraw yes, changed

**Review by realstraw — APPROVED** [review]

**Pull request merged**